hawkular-wildfly-agent/src/main/java/org/hawkular/agent/monitor/protocol/EndpointService.java

```diff
@@ -80,33 +82,49 @@
 
     private class InventoryListenerSupport {
         private final List<InventoryListener> inventoryListeners = new ArrayList<>();
+        private final ReadWriteLock inventoryListenerRWLock = new ReentrantReadWriteLock();
 
         public void fireResourcesAdded(List<Resource<L>> resources) {
             if (!resources.isEmpty()) {
-                LOG.debugf("Firing inventory event for [%d] added/modified resources", resources.size());
-                InventoryEvent<L> event = new InventoryEvent<L>(EndpointService.this, resources);
-                for (InventoryListener inventoryListener : inventoryListeners) {
-                    inventoryListener.resourcesAdded(event);
+                try {
+                    inventoryListenerRWLock.readLock().lock();
+                    LOG.debugf("Firing inventory event for [%d] added/modified resources", resources.size());
+                    InventoryEvent<L> event = new InventoryEvent<L>(EndpointService.this, resources);
+                    for (InventoryListener inventoryListener : inventoryListeners) {
+                        inventoryListener.resourcesAdded(event);
+                    }
+                } finally {
+                    inventoryListenerRWLock.readLock().unlock();
                 }
             }
         }
 
         public void fireResourcesRemoved(List<Resource<L>> resources) {
             if (!resources.isEmpty()) {
-                LOG.debugf("Firing inventory event for [%d] removed resources", resources.size());
-                InventoryEvent<L> event = new InventoryEvent<L>(EndpointService.this, resources);
-                for (InventoryListener inventoryListener : inventoryListeners) {
-                    inventoryListener.resourcesRemoved(event);
+                try {
+                    inventoryListenerRWLock.readLock().lock();
```

Comment:
acquisition of all locks should always come outside (just before) entering the try-catch. Because if something goes wrong, the finally block is going to try to unlock a lock you have not acquired. So swap lines 104-105. Do the same for the others.